bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -679,6 +679,32 @@ impl Connection {
         Ok(self)
     }
 
+    /// Serializes the session state from the connection.
+    ///
+    /// If no ticket or an invalid ticket is found, returns an empty buffer.
+    ///
+    /// Note: This function is not recommended for > TLS1.2 because in TLS1.3
+    /// servers can send multiple session tickets and this will return only
+    /// the most recently received ticket.
+    pub fn session(&mut self) -> Result<Vec<u8>, Error> {
+        let size;
+        unsafe {
+            size = s2n_connection_get_session_length(self.connection.as_ptr());
+        }
+        if size <= 0 {
+            return Ok(Vec::default());
+        }
+        let mut buf = vec![0; size as usize];
+        unsafe {
+            if s2n_connection_get_session(self.connection.as_ptr(), buf.as_mut_ptr(), size as usize)
+                != size
+            {
+                return Ok(Vec::default());
+            }
+        }
+        return Ok(buf);
+    }
```

Comment:
We prefer having the application pass a `&mut Vec<u8>`, in case they already have a destination to write to. Also, we should be using `into_result` rather than checking error codes manually.

I'd also prefer to call it `session_ticket` so it matches the `set_session_ticket` function above.

```suggestion
    pub fn session_ticket(&mut self, session: &mut Vec<u8>) -> Result<usize, Error> {
        let len = unsafe {
            s2n_connection_get_session_length(self.connection.as_ptr()).into_result()
        }? as usize;
        if len > 0 {
            session.reserve(len);
            unsafe {
                let buf = session.spare_capacity_mut();
                s2n_connection_get_session(self.connection.as_ptr(), buf.as_mut_ptr() as *mut _, len as usize).into_result()?;
                session.set_len(session.len() + len);
            }
        }
        return Ok(len);
    }
```

Reply:
> We prefer having the application pass a &mut Vec<u8>

I see both patterns in-use. example: https://github.com/aws/s2n-tls/blob/main/bindings/rust/s2n-tls/src/connection.rs#L968
 I agree that having client pass a buffer would be ideal, but there is no way for the client to know the correct size of the buffer beforehand (unless we also expose the session_length api) and having the length as a separate API did not feel idiomatic, thus went this route.


>  we should be using into_result rather than checking error codes manually.

Reply:
done